Generalize action testing framework

nilsdeppe opened the pull request:
## Proposed changes

Generalize action testing framework to:

- Have a local algorithm hold the DataBox (necessary for Actions that call a simple action)
- Replace some uses of `apply` with a function called `next_action` that evaluates only the next action in the action list. If the action is not ready and `ERROR` is hit.
- Replace remaining uses of `apply` with a function call `simple_action` that invokes the simple action on the component.

**Note:** Only commits the following commits are new:
- Replace apply with simple_action in mocking framework
- Add next_action, is_ready support to mocking framework
- Change returned DataBox type to rvalue in EventsAndTriggers
- Have mocking framework create all local algorithms
- Allow actions to retrieve component mocks
- Compute DataBox types in mocking framework
- Enclose ActionTesting_detail in ActionTesting namespace
- Move code around ActionRunner class

Depends on:
- [x] #912 

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Move code around ActionRunner class

- We will need some things to be public

nilsdeppe committed:
Enclose ActionTesting_detail in ActionTesting namespace

nilsdeppe committed:
Compute DataBox types in mocking framework

The MockLocalAlgorithm, which mocks AlgorithmImpl now holds a variant of
DataBoxes just like AlgorithmImpl, though they are not yet used.

nilsdeppe committed:
Allow actions to retrieve component mocks

When using the mocking framework it needs to be possible to retrieve a proxy to
a component that is being mocked through the non-mock component. That is, if I
am mocking the component A with MockA then if an action calls
Parallel::get_parallel_component<A>(cache) it should get back a proxy to
MockA. This commit enables such behavior.

nilsdeppe committed:
Have mocking framework create all local algorithms

The mocking framework for parallel components now requires users to create the
initial DataBox and explicitly insert the local algorithms into the mocking
code. In the future the user won't be able to pass any DataBoxes explicitly to
the `apply` function of the ActionRunner.

nilsdeppe committed:
Change returned DataBox type to rvalue in EventsAndTriggers

We no longer have a copy assignment in db::DataBox and the move assignment does
nothing when moving into oneself.

nilsdeppe committed:
Add next_action, is_ready support to mocking framework

The DataBox is stored on the MockLocalAlgorithm and the new next_action function
executes the next action in the action list on the MockLocalAlgorithm. This
follows exactly what the actual AlgorithmImpl class does. A few things to note:

- If the action isn't ready to be evaluated, a hard error is hit.
- An is_ready function is provided to allow manual checking of whether or not
  the next action is ready to be evaluated
- The apply function should only be used for simple actions (apply will be
  replaced soon)

nilsdeppe committed:
Replace apply with simple_action in mocking framework

wthrowe commented on tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/Test_FluxCommunicationLts.cpp:
```diff
@@ -0,0 +1,455 @@
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "DataStructures/Variables.hpp"
+// IWYU pragma: no_include "DataStructures/VariablesHelpers.hpp"  // for Variables
+
```
Remove blank line.

wthrowe commented on tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/Test_FluxCommunicationLts.cpp:
```diff
@@ -0,0 +1,455 @@
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
+#include "tests/Unit/ActionTesting.hpp"
+// #include "tests/Unit/TestHelpers.hpp"
```
Remove?

nilsdeppe replied:
Yea, IWYU experiment, forgot to delete...

wthrowe commented on tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/Test_FluxCommunicationLts.cpp:
```diff
@@ -0,0 +1,455 @@
+                  MortarRecorderTag, mortar_next_temporal_ids_tag<2>>,
+              typename my_component::compute_tags>(
+              self_step_start, self_step_end,
+
```
Remove blank line.

wthrowe commented on tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/Test_FluxCommunication.cpp:
```diff
@@ -434,51 +429,31 @@ SPECTRE_TEST_CASE("Unit.DiscontinuousGalerkin.Actions.FluxCommunication",
+  runner.next_action<my_component>(west_id);
+  CHECK(runner.is_ready<my_component>(self_id));
+
+  // DeceiveDataForFluxes
```
I know we're mocking, but...

nilsdeppe replied:
I really enjoyed this typo, thank you for noticing it :)

wthrowe reviewed (changes requested):
Travis OSX failures are real.

nilsdeppe committed:
Improve support for libc++

nilsdeppe commented:
I've done the following:
- Rebased onto the other PR (I'll do it again once that one gets merged)
- Added fixups addressing comments (thanks for catching all those!)
- Added a commit to allow better interoperability with libc++ (I was trying to test the macOS failure on my machine, hoping my libc++ version was close enough to reproduce, it wasn't)
- I've squashed a commit that hopefully fixes the macOS build errors. I believe they were due to libc++ not explicitly converting the ElementId to ElementIndex in the emplace call.

nilsdeppe commented:
Rebased on develop (@wthrowe can you verify the changes to Test_InitializeElement again? I want to make sure I didn't break anything that tests LTS :) )

kidder reviewed (dismissed):
Don't be surprised if I make suggestions after trying to use this framework...

nilsdeppe commented:
I'll be disappointed if you don't at least make documentation suggestions ;)

wthrowe reviewed (dismissed):
Test_InitializeElement looks fine.

nilsdeppe committed:
Bump macOS Xcode version to 9.2

nilsdeppe commented:
Okay, thanks for checking that again!

Squashed the commits.

nilsdeppe commented:
Travis failures are timeouts

kidder commented:
@wthrowe I think this is ready
